src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LinkedEditingRange/LinkedEditingRangeEndpoint.cs

```diff
@@ -0,0 +1,133 @@
+        // Internal for testing only.
+        internal readonly string _wordPattern = @"!?[^ <>!\/\?\[\]=""\\@" + Environment.NewLine + "]+";
+
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly TagHelperFactsService _tagHelperFactsService;
+
+        public LinkedEditingRangeEndpoint(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            TagHelperFactsService tagHelperFactsService)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _tagHelperFactsService = tagHelperFactsService;
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "linkedEditingRangeProvider";
+            var registrationOptions = new LinkedEditingRangeRegistrationOptions();
+            return new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
+        }
+
+        public async Task<LinkedEditingRanges> Handle(
+            LinkedEditingRangeParams request,
+            CancellationToken cancellationToken)
+        {
+            var document = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(
+                    request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken);
+
+            if (document is null || cancellationToken.IsCancellationRequested)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var location = await GetSourceLocation(request, document).ConfigureAwait(false);
+            var ancestors = GetAncestors(codeDocument, location);
+
+            // We only care if the user is within a TagHelper with a valid start and end tag.
+            if (_tagHelperFactsService.TryGetNearestAncestorStartAndEndTags(
+                    ancestors, out var startTag, out var endTag) &&
```

Comment:
nit: i'd roll this into the previous line. It doesn't breach our line length boundary does it?

Reply:
Personally I'd say 100 characters is too restrictive, but I don't feel strongly either way.

I don't think we have a defined maximum line length do we? I've certainly written some that were too long without an analyzer yelling at me.

Reply:
> Personally I'd say 100 characters is too restrictive, but I don't feel strongly either way.

+1, imo I typically go off of 150-200.

> I don't think we have a defined maximum line length do we? I've certainly written some that were too long without an analyzer yelling at me.

Ya I think we had it in the past but it was too short and didn't add a lot of value. That being said, if we had a length of like 200 or so (something kind of long) I think it could be a net positive.